### PR TITLE
Schematic for customisable error handling migration

### PR DESCRIPTION
**Pull request by zakhenry**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/ngrx/platform/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)

<strike>- [ ] Documentation has been added / updated (for bug fixes / features)</strike>

## PR Type

What kind of change does this PR introduce?

```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[x] Other... Please describe: Migration schematic to migrate breaking change introduced in effects.
```

## What is the current behavior?

https://github.com/ngrx/platform/pull/2295 introduces a breaking change, this PR adds a migration to ease the transition to ngrx@9 for developers.

Closes #

## What is the new behavior?

On upgrade to ngrx 9 developers can run the migration to automatically fix the renamed configuration variable

## Does this PR introduce a breaking change?

```
[ ] Yes
[x] No
```

## Other information
~This PR is pending the merge of https://github.com/ngrx/platform/pull/2295, thus it currently contains those commits.~

### CODE REVIEW

File: projects/ngrx.io/content/guide/effects/lifecycle.md [hunk 1]
```diff
@@ -24,7 +24,7 @@ Usage:
 
 <code-example header="log.effects.ts">
 import { Injectable } from '@angular/core';
-import { Actions, Effect, ofType } from '@ngrx/effects';
+import { Actions, createEffect, ofType } from '@ngrx/effects';
```
Comment (zakhenry) [inline projects/ngrx.io/content/guide/effects/lifecycle.md]: this was just a driveby change that I noticed the import was wrong

File: modules/effects/migrations/9_0_0/index.ts [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+) {
+  const changes: ReplaceChange[] = [];
+  ts.forEachChild(sourceFile, node => findKey(node, changes));
+  return changes;
+
+  function findKey(node: ts.Node, changes: ReplaceChange[]) {
+    if (
+      ts.isIdentifier(node) &&
+      node.text === OLD_TEXT &&
+      isEffectDecorator(node.parent.parent.parent.parent)
```
Comment (timdeschryver) [inline modules/effects/migrations/9_0_0/index.ts]: I'm worried that this (`parent.parent.parent`) might throw an error in some cases.
What about finding the `createEffect` first, and then looking for `resubscribeOnError`?

File: modules/effects/migrations/9_0_0/index.ts [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+) {
+  const changes: ReplaceChange[] = [];
+  ts.forEachChild(sourceFile, node => findKey(node, changes));
+  return changes;
+
+  function findKey(node: ts.Node, changes: ReplaceChange[]) {
+    if (
+      ts.isIdentifier(node) &&
+      node.text === OLD_TEXT &&
+      isCreateEffectFunction(node.parent.parent.parent)
```
Comment (timdeschryver) [inline modules/effects/migrations/9_0_0/index.ts]: Same here

File: modules/effects/migrations/9_0_0/index.ts (hunk 2 above)
Comment (zakhenry) [reply modules/effects/migrations/9_0_0/index.ts]: Hm I worried about this too, so tested it. Interestingly node.parent is never undefined, I think it just becomes recursive at the top level somehow. I agree though that it would be less confusing to find the effect context first and then descend into children so I'll change that 👍 

It will be a little more complex I suspect as I will need to make sure not to descend into the effect declaration as that could plausibly contain identifiers that I shouldn't be changing.

Comment (zakhenry) [reply modules/effects/migrations/9_0_0/index.ts]: Ok refactor done, I actually went with a different strategy (visitor pattern) which ended up being a lot cleaner

File: modules/effects/migrations/9_0_0/index.ts (hunk 3 above)
Comment (zakhenry) [reply modules/effects/migrations/9_0_0/index.ts]: done

**Comment by zakhenry** [comment]
Rebased on master now that the dependent PR has merged.

File: modules/effects/migrations/9_0_0/index.ts [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+  sourceFile: ts.SourceFile,
+  oldText: string,
+  newText: string
+): ReplaceChange[] {
+  const changes: ReplaceChange[] = [];
+
+  ts.forEachChild(sourceFile, node => {
+    visitCreateEffectFunctionCreator(node, createEffectNode => {
+      // only search the second arg to createEffect() (the config object)
+      if (createEffectNode.arguments[1]) {
```
Comment (timdeschryver) [inline modules/effects/migrations/9_0_0/index.ts]: Personally I would rather see this as:

```ts
const [configNode] = createEffectNode.arguments
```

This also makes the comment obsolete

File: modules/effects/migrations/9_0_0/index.spec.ts [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+        {
+          "include": [**./*.ts"]
+        }
+       `
+    );
+    createPackageJson('', pkgName, appTree);
+  });
+
+  describe('Replaces resubscribeOnError with useEffectsErrorHandler in effect options', () => {
+    describe('should replace resubscribeOnError configuration key with useEffectsErrorHandler', () => {
```
Comment (timdeschryver) [inline modules/effects/migrations/9_0_0/index.spec.ts]: this describe can be left out? It has more or less the same description as the parent

File: modules/effects/migrations/9_0_0/index.spec.ts [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+      this.actions$.pipe(
+        tap(action => console.log(action))
+      ), { useEffectsErrorHandler: false });
+  }
+        `;
+
+        test(input, expected);
+      });
+
+      it('in @Effect() effect creator', () => {
```
Comment (timdeschryver) [inline modules/effects/migrations/9_0_0/index.spec.ts]:
```suggestion
      it('in @Effect() decorator', () => {
```

File: projects/ngrx.io/content/guide/effects/lifecycle.md (hunk 1 above)
Comment (timdeschryver) [reply projects/ngrx.io/content/guide/effects/lifecycle.md]: Could you create a separate PR for this change please?

**Comment by timdeschryver** [comment]
I like this approach more in comparison to what we first had 👏

File: projects/ngrx.io/content/guide/effects/lifecycle.md (hunk 1 above)
Comment (zakhenry) [reply projects/ngrx.io/content/guide/effects/lifecycle.md]: sure

File: modules/effects/migrations/9_0_0/index.ts (hunk 4 above)
Comment (zakhenry) [reply modules/effects/migrations/9_0_0/index.ts]: yep done

File: modules/effects/migrations/9_0_0/index.spec.ts (hunk 5 above)
Comment (zakhenry) [reply modules/effects/migrations/9_0_0/index.spec.ts]: ah actually you've picked up a mistake in my describe nesting. the structure should be

```
Effects Migration 9_0_0
└── Replaces resubscribeOnError with useEffectsErrorHandler in effect options
    ├── should replace resubscribeOnError configuration key with useEffectsErrorHandler
    |   ├── in createEffect() effect creator
    |   ├── in @Effect() decorator
    └── should not replace non-ngrx identifiers
        ├── in module scope
        └── within create effect callback
```

fixed.

File: modules/effects/migrations/9_0_0/index.spec.ts (hunk 6 above)
Comment (zakhenry) [reply modules/effects/migrations/9_0_0/index.spec.ts]: done

File: projects/ngrx.io/content/guide/effects/lifecycle.md (hunk 1 above)
Comment (zakhenry) [reply projects/ngrx.io/content/guide/effects/lifecycle.md]: moved to https://github.com/ngrx/platform/pull/2336
